source/extensions/transport_sockets/tls/context_impl.cc

```diff
@@ -506,16 +616,39 @@ int ContextImpl::ignoreCertificateExpirationCallback(int ok, X509_STORE_CTX* ctx
   return ok;
 }
 
-int ContextImpl::verifyCallback(X509_STORE_CTX* store_ctx, void* arg) {
-  ContextImpl* impl = reinterpret_cast<ContextImpl*>(arg);
-  SSL* ssl = reinterpret_cast<SSL*>(
-      X509_STORE_CTX_get_ex_data(store_ctx, SSL_get_ex_data_X509_STORE_CTX_idx()));
+ssl_verify_result_t ContextImpl::verifyCallback(SSL* ssl, uint8_t* out_alert) {
+  ContextImpl* impl = reinterpret_cast<ContextImpl*>(SSL_CTX_get_app_data(SSL_get_SSL_CTX(ssl)));
   Envoy::Ssl::SslExtendedSocketInfo* sslExtendedInfo =
       reinterpret_cast<Envoy::Ssl::SslExtendedSocketInfo*>(
           SSL_get_ex_data(ssl, ContextImpl::sslExtendedSocketInfoIndex()));
 
+  const STACK_OF(CRYPTO_BUFFER)* buffers = SSL_get0_peer_certificates(ssl);
+  bssl::UniquePtr<X509> cert = nullptr;
+  bssl::UniquePtr<STACK_OF(X509)> chain(sk_X509_new(nullptr));
+  for (size_t i = 0; i < sk_CRYPTO_BUFFER_num(buffers); ++i) {
+    bssl::UniquePtr<X509> x509(X509_parse_from_buffer(sk_CRYPTO_BUFFER_value(buffers, i)));
+    if (!x509) {
+      continue;
```

Comment:
What causes this case to happen? Should this be a failure? Please add comment.

Reply:
When peer sends corrupted certificate. Not sure if this can really happen (i.e. if it wouldn't be rejected sooner in the stack), but changed to failure. Good catch, thanks!